### DISCUSSION TITLE
[Translator] How about a built in event for Locale definition?

### DISCUSSION THREAD

**Discussion by Nayte91** [symfony/symfony#50304] [Q&A]
Hello,
I was playing with symfony translator recently, and spent some time to prepare my app to change locale based on browser's default, based on a user's chosen language in a dropdown menu and registered on DB, based on Kernel events.

While the new LocaleSwitcher tool is a blast, I felt like the need to create 2 listeners (LocaleListener and UserLocaleListener) to handle all the needs can be optimized.

So here's my question: I was wondering if it's possible to add a built in event for the translator component, that will be fired, for example, between the KernelEvents::REQUEST and the SecurityEvents::INTERACTIVE_LOGIN? Or Maybe after the SecurityEvents? Is it a thing with Symfony to add events during the request lifetime like this? I think that's what happen when installing the security component or doctrine, they ship with their very own events that we can plug on.

If so, ca we pack in a LocaleDefinitionEvent class all the tools needed to take decision?
Request currentRequest, 
null|UserInterface currentUser,
Session,
LocalSwitcher

As I can whitness, it seems we fire some actions during the KernelEvent::REQUEST event because we need to do it before the controller's route is defined. And we plug to INTERACTIVE_LOGIN event because it gives the current logged User object to play with. 

- The final goal is to have only one listener, search what is the browser's value or if any logged user preference is written somewhere.
- Bonus happy ending can be to automatically define locale based on user or browser, e.g. by telling to a config key in translation.yaml what is the user's property that carry this information (example a key user_locale: foo, because I register in User::foo property which locale to use like fr, de, jp, ...)

As I'm pretty new but I want to start contributing, I can handle this if someone tell me that I'm not totally wrong with how things work/how we can deal with this. I'm aware that's not an issue nor a bug, more like DX improvement/sexy QoL improvement.

If this idea is quite OK, can some experienced Symfonist give me links to related documentation, as I can work on it?

Thank you for reading me!

**Comment by MatTheCat** [comment]
You can already use the special `_locale` routing parameter or `framework.set_locale_from_accept_language` (the former overrides the latter) to set the locale from the request.

How to set it from the logged user is documented in https://symfony.com/doc/current/translation.html#handling-the-user-s-locale

If these do address your issue another event shouldn’t be needed!
